llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp

```diff
@@ -3185,10 +3205,15 @@ NVPTXTargetLowering::LowerSTOREVector(SDValue Op, SelectionDAG &DAG) const {
     case MVT::v4f16:
     case MVT::v4bf16:
     case MVT::v4f32:
-    case MVT::v8f16: // <4 x f16x2>
+      // This is a "native" vector type
+      break;
+    case MVT::v8i8:   // <2 x i8x4>
+    case MVT::v8f16:  // <4 x f16x2>
     case MVT::v8bf16: // <4 x bf16x2>
     case MVT::v8i16:  // <4 x i16x2>
-      // This is a "native" vector type
+    case MVT::v16i8:  // <4 x i8x4>
+      // This can be upsized into a "native" vector type
+      UpsizeElementTypes = true;
```

Comment:
This whole switch should probably be folded into `GetUpsizedNumEltsAndEltVT` and `UpsizeElementTypes` returned as an additional `bool`.
Then we could just use `auto [Upsize, NumElts, EltVT] = GetUpsizedNumEltsAndEltVT(ValVT.getVectorNumElements(), ValVT.getVectorElementType())`

Reply:
Hmm, do you think that will hurt readability? I think having the Upsize helper compute NumElts and EltVT for all (including non-upsized) cases is a bit more confusing.

Reply:
I look at it as "Calculate the right vector shape for lowering", where the "right shape" may be identical to the input shape.

Reply:
Got it. That sounds reasonable, although we would also need to return a "Native" boolean, right? Returning early if it is false?

Reply:
Or it could be derived from EltVT != original EltVT.

Reply:
I think there's a misunderstanding; I was referring to cases that are neither native _nor_ upsize-able. This switch statement has two functions right now, marking the cases that we need to upsize, and returning early for cases that cannot be handled at all, such as `v8i32`. How will we handle the latter with the helper function approach, besides making it a "Try" function and adding another boolean to the returns?

Reply:
I implemented the above std::tie solution but let me know if you still think there is a better way. I'm still hesitant to fold the entire switch into the helper function because the switch is responsible for exiting early from `ReplaceLoadVector`/`LowerSTOREVector` and I don't think there is a clean way to do it, but if I'm missing something I'm happy to tweak it.

Reply:
Fair enough. So, the answers we're looking for are: should we handle this vector, and if so, how.

That asks for either returning a `std::optional<std::pair<>>` or just represent "can't do it" as `NumElts == 0`.
The latter is probably more concise. The function entry would look roughly like this:
```
auto NumElts, EltType = getVectorLoweringParams(Val);
if (NumElts == 0)
  return SDValue();

... lower as `NumElts x EltType`.
```

Reply:
std::optional is fine, too.

Reply:
Great, I just pushed the implementation, let me know what you think.